File: python3/shortest_unsorted_continuous_subarray.py
"""
Space   : O(1)
Time    : O(n)
"""


class Solution:
    def findUnsortedSubarray(self, nums: List[int]) -> int:
        n = len(nums)
        start, end = n-1, 0

        isAsc = True
        forward = nums[0]
        for i in range(1, n):
            if nums[i] < forward:
                end = max(end, i)
                isAsc = False
            forward = max(forward, nums[i])

        if isAsc:
            return 0

        backward = nums[-1]
        for j in range(n-2, -1, -1):
            if nums[j] > backward:
                start = min(start, j)
            backward = min(backward, nums[j])

        return end - start + 1

File: golang/shortest_unsorted_continuous_subarray.go
package main

// Space   : O(1)
// Time    : O(n)

func findUnsortedSubarray(nums []int) int {
	n := len(nums)
	start, end := n-1, 0

	isAsc := true
	high := nums[0]
	for i := 1; i < n; i++ {
		if nums[i] < high {
			end = Max(end, i)
			isAsc = false
		}
		high = Max(high, nums[i])
	}

	if isAsc {
		return 0
	}

	low := nums[n-1]
	for j := n - 2; j >= 0; j-- {
		if nums[j] > low {
			start = Min(start, j)
		}
		low = Min(low, nums[j])
	}

	return end - start + 1
}
